### ISSUE TITLE
@Data and @Builder don't work together or with constructor annotations

### ISSUE THREAD

**Issue by antillean** [rzwitserloot/lombok#1912]
## Problem

From reading the documentation, I expected that I'd be able to do this

```java
@Data
@Builder
@NoArgsConstructor
class Point {
...
}
```

and get a class with getters, setters and a builder, and with the following constructors:

1. Required args constructor (from `@Data`)
2. All args constructor (from `@Builder`)
3. No args constructor (from `@NoArgsConstructor`).

But I don't. Instead I get a compilation error like this:

```
./gradlew clean build

> Task :compileJava FAILED
/path/to/workspace/lombok-bug/src/main/java/Point.java:7: error: constructor Point in class Point cannot be applied to given types;
@Builder
^
  required: no arguments
  found: Double,Double,Double
  reason: actual and formal argument lists differ in length
1 error
```

## Reproduction

To reproduce this issue, follow the instructions in [this repo](https://github.com/antillean/lombok-constructor-bug)'s readme (which are just to clone it and build it).

That repo contains an example using a simple point class and some tests for that class. This is the point class.

```java
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.NonNull;

@Data
@Builder
@NoArgsConstructor
class Point {
  @NonNull private Double x;
  @NonNull private Double y;
  private Double z;
}
```

## Affected versions

I tested v1.18.2, v1.18.0 and v1.16.22. They were all affected by this issue.

## Some digging

Here are some things I tried, what I expected, and what I got.

| Annotations | Expected Constructors | Actual Constructors | Works |
| - | - | - | - |
| `@Data`<br>`@NoArgsConstructor` | Required args<br>No args | No args |  :-1:  |
| `@Data`<br> `@Builder` | Required args<br>All args | All args |  :-1:  |
| `@Data`<br> `@Builder`<br> `@NoArgsConstructor` | Required args <br> All args <br> No args | No args |  :-1: |
| `@NoArgsConstructor`<br> `@AllArgsConstructor`<br> `@RequiredArgsConstructor` | No args<br>All args<br>Required args | No args<br>All args<br>Required args | :+1: |

So the problem only seems to happen when either `@Data` or `@Builder` is combined with one of the args constructors, or when `@Data` and `@Builder` are used together.

(This meant that, for my original situation which should've only required `@Data @Builder @NoArgsConstructor` to get Lombok and Jackson to play nicely together, I needed `@Data @Builder @NoArgsConstructor @AllArgsConstructor @RequiredArgsConstructor`.)

So basically it seems that:

1. `@Data`'s constructor is ignored when `@Builder` is also present.
2. Both `@Data`'s and `@Builder`'s constructors are ignored if any explicit constructor annotation is present.
3. Explicit constructor annotations work together as expected.

**Comment by janinko** [comment]
I would add that lombok documentation and javadoc differs.

[Builder javadoc](https://projectlombok.org/api/lombok/Builder.html) reads:
> If a class is annotated, then a private constructor is generated with all fields as arguments (as if `@AllArgsConstructor(AccessLevel.PRIVATE)` is present on the class), and it is as if this constructor has been annotated with `@Builder` instead.

[Builder documentaion](https://projectlombok.org/features/Builder) reads:
> Finally, applying `@Builder` to a class is as if you added `@AllArgsConstructor(access = AccessLevel.PACKAGE)` to the class and applied the `@Builder` annotation to this all-args-constructor. This only works if you haven't written any explicit constructors yourself.

Please clearly specify what behaviour is correct and possibly fix the behaviour to reflect that.

**Comment by rzwitserloot** [comment]
Lombok's current behaviour is intentional; add an `@AllArgsConstructor` to the mix to get what you want.

Updated the javadoc to reflect this.

**Comment by lennarth-anaya** [comment]
I would like my client developers to use builder, while some DI frameworks demand constructors, which I really don't care since I don't see the code. This rule is in my opinion unnecessary.

I just added @AllArgsConstructor and still @Builder fails.

**Comment by Maaartinus** [comment]
@lennarth-anaya 

> I would like my client developers to use builder, while some DI frameworks demand constructors, which I really don't care since I don't see the code. This rule is in my opinion unnecessary.

Which rule? The OP stated three of them. The first one is questionable: Should `@Data`'s constructor win or should `@Builder`'s constructor win? IMHO, both should get generated, but I guess, this may lead to problems, sometimes. The other two rules are rather obvious. What do you mean?

> I just added `@AllArgsConstructor` and still `@Builder` fails.

I'm afraid, nobody can help you until you say what exactly happens in your case.

**Comment by lennarth-anaya** [comment]
I'm sorry for the delay, I read it on my mobile but forgot to answer on my desktop.

It seems another unrelated compiling issue was causing Lombok not to do its stuff because once all other compiling issues were resolved, lombok ceased bothering me with this AllArgConstructor/Builder missing issue. I can delete my comments if this is not useful to anyone else.

**Comment by dennisvang** [comment]
> Lombok's current behaviour is intentional; add an `@AllArgsConstructor` to the mix to get what you want.

@rzwitserloot is the [described behaviour] for `@Data` + `@NoArgsConstructor` (without `@Builder`) also intentional?

That surprises me, due to the following text from the [`@NoArgsConstructor` docs]:

>`@NoArgsConstructor` will generate a constructor with no parameters. [...] This annotation is useful primarily in combination with either `@Data` or one of the other constructor generating annotations. 

From the [`@Data` docs]:

>`@Data` is like having [...] and `@RequiredArgsConstructor` annotations on the class (except that no constructor will be generated if any explicitly written constructors already exist).

Does `@NoArgsConstructor` count as *"any explicitly written constructors"*? (I would not expect that)

Actually, with the following (contrived) class, I'm seeing generated getter, setter, equals, and so on, but *no* constructors at all (not even the no args constructor mentioned in the OP):

```java
@Data
@NoArgsConstructor
public class User {
    @NonNull
    private String name;
}

```

Just for verification: If I remove the  `@NoArgsConstructor`, I do see the expected required args constructor.

lombok version: 1.18.40

[`@Data` docs]: https://projectlombok.org/features/Data
[`@NoArgsConstructor` docs]: https://projectlombok.org/features/constructor
[described behaviour]: https://github.com/projectlombok/lombok/issues/1912#issue-371912942
